pkg/registry/graph.go

```diff
@@ -0,0 +1,176 @@
+package registry
+
+import (
+	"fmt"
+
+	"github.com/openshift/ci-tools/pkg/api"
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+// Component contains the name of a registry component and pointers for
+// either a reference, chain, or workflow
```

Comment:
We know we have strong one-way relationships (a child must be a chain or ref, a parent must be a chain or workflow and workflows must be roots) so I think we could do better with typing here.

Reply:
We need to have a quick way to access any node by name if we want to be able to get all parents/descendants quickly, and being able to put any of these into a single type that can be put into a map makes this simple. Otherwise, we would need to have 3 maps again, one for each of refs, chains, and workflows to be able to find the nodes we want quickly.

One thing we could do is just have the name and an enum for the type instead of 3 pointers. This would simplify the `Component` type, but would require us to have access to the refs, chains, and workflow maps if we need to access the contents of one of the components (vs the current type, which just points to the contents directly).

Reply:
I think specifically my feedback here is really similar to the internal/external types. We are using a strongly typed language and all we do by fudging these sorts of declarations is we make possible in code strictly forbidden combinations. This can lead to programmer bugs and frustration when implicitly invalid combinations either happen in practice, become explicitly possible and need refactoring, etc. If you think that there is no simple alternative to this type, I think it's reasonable to fudge it but in my opinion I'd prefer to invest some nontrivial amount of effort creating a tight type system that doesn't let us mess something up down the road. Of course if the upfront development cost is great, then it's a crapshoot.